File: src/service/ServiceManager.js
const EventEmitter = require('events');
const { v4: uuidv4 } = require('uuid');
const ServiceRegistry = require('./ServiceRegistry');
const HealthCheck = require('./HealthCheck');
const ConfigManager = require('./ConfigManager');
const Logger = require('../monitor/Logger');

/**
 * ServiceManager - 服务管理器
 * 负责服务的生命周期管理、配置、监控等
 */
class ServiceManager extends EventEmitter {
  constructor(options = {}) {
    super();
    this.logger = new Logger();
    this.services = new Map();
    this.registry = new ServiceRegistry();
    this.healthCheck = new HealthCheck();
    this.configManager = new ConfigManager();
    
    this.options = {
      checkInterval: options.checkInterval || 30000,
      maxRetries: options.maxRetries || 3,
      retryDelay: options.retryDelay || 1000,
      timeout: options.timeout || 5000
    };
    
    this.checkTimer = null;
  }
  
  /**
   * 创建服务
   */
  async createService(serviceConfig) {
    const serviceId = serviceConfig.id || uuidv4();
    
    // 验证配置
    await this.configManager.validate(serviceConfig);
    
    const service = {
      id: serviceId,
      name: serviceConfig.name,
      type: serviceConfig.type,
      status: 'created',
      config: serviceConfig,
      createdAt: new Date(),
      updatedAt: new Date(),
      metrics: {
        requests: 0,
        errors: 0,
        responseTime: 0
      }
    };
    
    this.services.set(serviceId, service);
    this.registry.register(serviceId, service);
    
    this.logger.info({ message: 'Service created', serviceId, name: service.name });
    this.emit('service:created', service);
    
    return service;
  }
  
  /**
   * 启动服务
   */
  async startService(serviceId) {
    const service = this.services.get(serviceId);
    if (!service) {
      throw new Error(`Service ${serviceId} not found`);
    }
    
    if (service.status === 'running') {
      return service;
    }
    
    try {
      service.status = 'starting';
      this.emit('service:starting', { serviceId });
      
      // 模拟启动过程
      await this.executeStartup(service);
      
      service.status = 'running';
      service.startedAt = new Date();
      service.updatedAt = new Date();
      
      this.logger.info({ message: 'Service started', serviceId });
      this.emit('service:started', service);
      
      return service;
    } catch (error) {
      service.status = 'failed';
      this.logger.error({ message: 'Failed to start service', serviceId, error: error.message });
      this.emit('service:failed', { serviceId, error: error.message });
      throw error;
    }
  }
  
  /**
   * 停止服务
   */
  async stopService(serviceId) {
    const service = this.services.get(serviceId);
    if (!service) {
      throw new Error(`Service ${serviceId} not found`);
    }
    
    if (service.status === 'stopped') {
      return service;
    }
    
    try {
      service.status = 'stopping';
      this.emit('service:stopping', { serviceId });
      
      // 模拟关闭过程
      await this.executeShutdown(service);
      
      service.status = 'stopped';
      service.stoppedAt = new Date();
      service.updatedAt = new Date();
      
      this.logger.info({ message: 'Service stopped', serviceId });
      this.emit('service:stopped', service);
      
      return service;
    } catch (error) {
      this.logger.error({ message: 'Failed to stop service', serviceId, error: error.message });
      throw error;
    }
  }
  
  /**
   * 重启服务
   */
  async restartService(serviceId) {
    await this.stopService(serviceId);
    return await this.startService(serviceId);
  }
  
  /**
   * 更新服务配置
   */
  async updateServiceConfig(serviceId, newConfig) {
    const service = this.services.get(serviceId);
    if (!service) {
      throw new Error(`Service ${serviceId} not found`);
    }
    
    // 验证新配置
    await this.configManager.validate(newConfig);
    
    const oldConfig = service.config;
    service.config = { ...oldConfig, ...newConfig };
    service.updatedAt = new Date();
    
    // 如果服务运行中，需要重新加载配置
    if (service.status === 'running') {
      await this.reloadServiceConfig(serviceId);
    }
    
    this.logger.info({ message: 'Service config updated', serviceId });
    this.emit('service:configUpdated', service);
    
    return service;
  }
  
  /**
   * 获取服务
   */
  getService(serviceId) {
    return this.services.get(serviceId);
  }
  
  /**
   * 获取所有服务
   */
  getAllServices() {
    return Array.from(this.services.values());
  }
  
  /**
   * 删除服务
   */
  async deleteService(serviceId) {
    const service = this.services.get(serviceId);
    if (!service) {
      throw new Error(`Service ${serviceId} not found`);
    }
    
    if (service.status === 'running') {
      await this.stopService(serviceId);
    }
    
    this.services.delete(serviceId);
    this.registry.unregister(serviceId);
    
    this.logger.info({ message: 'Service deleted', serviceId });
    this.emit('service:deleted', { serviceId });
  }
  
  /**
   * 启动健康检查
   */
  async startHealthCheck() {
    this.checkTimer = setInterval(async () => {
      for (const [serviceId, service] of this.services.entries()) {
        if (service.status === 'running') {
          const isHealthy = await this.healthCheck.check(service);
          
          if (!isHealthy && service.retries < this.options.maxRetries) {
            service.retries = (service.retries || 0) + 1;
            this.logger.warn({ message: 'Service unhealthy', serviceId, retry: service.retries });
            
            try {
              await this.restartService(serviceId);
            } catch (error) {
              this.logger.error({ message: 'Failed to restart service', serviceId });
            }
          }
        }
      }
    }, this.options.checkInterval);
    
    this.logger.info({ message: 'Health check started', interval: this.options.checkInterval });
  }
  
  /**
   * 执行启动流程（模拟）
   */
  async executeStartup(service) {
    return new Promise((resolve) => {
      setTimeout(() => {
        service.retries = 0;
        resolve();
      }, 100);
    });
  }
  
  /**
   * 执行关闭流程（模拟）
   */
  async executeShutdown(service) {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve();
      }, 100);
    });
  }
  
  /**
   * 重新加载服务配置
   */
  async reloadServiceConfig(serviceId) {
    // 实现配置重新加载逻辑
    return true;
  }
  
  /**
   * 关闭服务管理器
   */
  async shutdown() {
    if (this.checkTimer) {
      clearInterval(this.checkTimer);
    }
    
    // 停止所有运行中的服务
    for (const [serviceId, service] of this.services.entries()) {
      if (service.status === 'running') {
        await this.stopService(serviceId).catch(err => {
          this.logger.error({ message: 'Error stopping service during shutdown', serviceId });
        });
      }
    }
    
    this.logger.info({ message: 'ServiceManager shutdown' });
  }
}

module.exports = ServiceManager;

File: src/main/java/com/mapengine/service/ServiceManager.java
package com.mapengine.service;

import lombok.extern.slf4j.Slf4j;

import java.util.*;
import java.util.concurrent.*;
import java.util.function.Consumer;

@Slf4j
public class ServiceManager {
    private final Map<String, Service> services = new ConcurrentHashMap<>();
    private final ServiceRegistry registry = new ServiceRegistry();
    private final HealthCheck healthCheck = new HealthCheck();
    private final ConfigManager configManager = new ConfigManager();

    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
        Thread t = new Thread(r, "service-health-check");
        t.setDaemon(true);
        return t;
    });

    private final long checkIntervalMillis;
    private final int maxRetries;

    public ServiceManager() {
        this(30_000L, 3);
    }

    public ServiceManager(long checkIntervalMillis, int maxRetries) {
        this.checkIntervalMillis = checkIntervalMillis;
        this.maxRetries = maxRetries;
    }

    public Service createService(Map<String, Object> serviceConfig) throws Exception {
        configManager.validate(serviceConfig);
        String id = serviceConfig.containsKey("id") ? String.valueOf(serviceConfig.get("id")) : UUID.randomUUID().toString();
        String name = (String) serviceConfig.getOrDefault("name", "service-" + id);
        String type = (String) serviceConfig.getOrDefault("type", "generic");
        Service s = new Service(id, name, type, serviceConfig);
        services.put(id, s);
        registry.register(id, s);
        log.info("Service created: {} ({})", id, name);
        return s;
    }

    public Service startService(String serviceId) throws Exception {
        Service s = services.get(serviceId);
        if (s == null) throw new NoSuchElementException("Service " + serviceId + " not found");
        if ("running".equals(s.getStatus())) return s;
        s.setStatus("starting");
        try {
            executeStartup(s);
            s.setStatus("running");
            s.setStartedAt(java.time.Instant.now());
            s.setUpdatedAt(java.time.Instant.now());
            log.info("Service started: {}", serviceId);
            return s;
        } catch (Exception e) {
            s.setStatus("failed");
            log.error("Failed to start service {}", serviceId, e);
            throw e;
        }
    }

    public Service stopService(String serviceId) throws Exception {
        Service s = services.get(serviceId);
        if (s == null) throw new NoSuchElementException("Service " + serviceId + " not found");
        if ("stopped".equals(s.getStatus())) return s;
        s.setStatus("stopping");
        try {
            executeShutdown(s);
            s.setStatus("stopped");
            s.setStoppedAt(java.time.Instant.now());
            s.setUpdatedAt(java.time.Instant.now());
            log.info("Service stopped: {}", serviceId);
            return s;
        } catch (Exception e) {
            log.error("Failed to stop service {}", serviceId, e);
            throw e;
        }
    }

    public Service restartService(String serviceId) throws Exception {
        stopService(serviceId);
        return startService(serviceId);
    }

    public Service updateServiceConfig(String serviceId, Map<String, Object> newConfig) throws Exception {
        Service s = services.get(serviceId);
        if (s == null) throw new NoSuchElementException("Service " + serviceId + " not found");
        configManager.validate(newConfig);
        s.getConfig().putAll(newConfig);
        s.setUpdatedAt(java.time.Instant.now());
        log.info("Service config updated: {}", serviceId);
        // if running, reload config (noop default)
        if ("running".equals(s.getStatus())) reloadServiceConfig(serviceId);
        return s;
    }

    public Service getService(String serviceId) {
        return services.get(serviceId);
    }

    public List<Service> getAllServices() {
        return new ArrayList<>(services.values());
    }

    public void deleteService(String serviceId) throws Exception {
        Service s = services.get(serviceId);
        if (s == null) throw new NoSuchElementException("Service " + serviceId + " not found");
        if ("running".equals(s.getStatus())) stopService(serviceId);
        services.remove(serviceId);
        registry.unregister(serviceId);
        log.info("Service deleted: {}", serviceId);
    }

    public void startHealthChecks() {
        scheduler.scheduleAtFixedRate(() -> {
            for (Service s : services.values()) {
                if ("running".equals(s.getStatus())) {
                    try {
                        boolean healthy = healthCheck.check(s);
                        if (!healthy) {
                            int retries = (int) s.getConfig().getOrDefault("retries", 0);
                            if (retries < maxRetries) {
                                s.getConfig().put("retries", retries + 1);
                                log.warn("Service unhealthy, attempting restart: {} (retry={})", s.getId(), retries + 1);
                                try { restartService(s.getId()); } catch (Exception ex) { log.error("Restart failed for {}", s.getId(), ex); }
                            } else {
                                log.warn("Service {} exceeded max retries", s.getId());
                            }
                        }
                    } catch (Exception e) {
                        log.error("Health check failed for {}", s.getId(), e);
                    }
                }
            }
        }, checkIntervalMillis, checkIntervalMillis, TimeUnit.MILLISECONDS);
        log.info("Health checks scheduled every {} ms", checkIntervalMillis);
    }

    public void stopHealthChecks() {
        scheduler.shutdownNow();
        log.info("Health checks stopped");
    }

    protected void executeStartup(Service s) throws InterruptedException {
        // simulate startup delay
        Thread.sleep(100);
        s.setConfig(s.getConfig() == null ? new ConcurrentHashMap<>() : s.getConfig());
        s.getConfig().put("retries", 0);
    }

    protected void executeShutdown(Service s) throws InterruptedException {
        Thread.sleep(100);
    }

    protected void reloadServiceConfig(String serviceId) {
        // noop default - override as needed
    }

    public ServiceRegistry getRegistry() { return registry; }
}
